### PR TITLE
Optimize TrimGraphQLTypes and To[Camel/Pascal]Case

### PR DESCRIPTION
**Pull request by sungam3r**
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.16299.726 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7 CPU 920 2.67GHz (Nehalem), 1 CPU, 8 logical and 4 physical cores
Frequency=2607421 Hz, Resolution=383.5207 ns, Timer=TSC
.NET Core SDK=5.0.101
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  DefaultJob : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT


```
|              Method |                 Name |         Mean |     Error |    StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------- |--------------------- |-------------:|----------:|----------:|------:|-------:|------:|------:|----------:|
| **TrimGraphQLTypesOld** |                     **** |    **100.82 ns** |  **0.315 ns** |  **0.263 ns** |  **1.00** |      **-** |     **-** |     **-** |         **-** |
| TrimGraphQLTypesNew |                      |     18.98 ns |  0.008 ns |  0.007 ns |  0.19 |      - |     - |     - |         - |
|                     |                      |              |           |           |       |        |       |       |           |
| **TrimGraphQLTypesOld** |                **Human** |    **192.50 ns** |  **0.343 ns** |  **0.304 ns** |  **1.00** |      **-** |     **-** |     **-** |         **-** |
| TrimGraphQLTypesNew |                Human |    103.35 ns |  0.026 ns |  0.022 ns |  0.54 |      - |     - |     - |         - |
|                     |                      |              |           |           |       |        |       |       |           |
| **TrimGraphQLTypesOld** |               **Human!** |    **502.12 ns** |  **2.092 ns** |  **1.854 ns** |  **1.00** | **0.0572** |     **-** |     **-** |     **240 B** |
| TrimGraphQLTypesNew |               Human! |    138.39 ns |  0.270 ns |  0.252 ns |  0.28 | 0.0076 |     - |     - |      32 B |
|                     |                      |              |           |           |       |        |       |       |           |
| **TrimGraphQLTypesOld** |              **[Human]** |    **795.21 ns** |  **2.356 ns** |  **2.204 ns** |  **1.00** | **0.1068** |     **-** |     **-** |     **448 B** |
| TrimGraphQLTypesNew |              [Human] |    156.77 ns |  0.188 ns |  0.167 ns |  0.20 | 0.0076 |     - |     - |      32 B |
|                     |                      |              |           |           |       |        |       |       |           |
| **TrimGraphQLTypesOld** |             **[Human]!** |  **1,077.61 ns** |  **2.239 ns** |  **2.094 ns** |  **1.00** | **0.1564** |     **-** |     **-** |     **656 B** |
| TrimGraphQLTypesNew |             [Human]! |    170.84 ns |  0.411 ns |  0.364 ns |  0.16 | 0.0076 |     - |     - |      32 B |
|                     |                      |              |           |           |       |        |       |       |           |
| **TrimGraphQLTypesOld** |         **[[Human!]!]!** |  **2,170.29 ns** |  **3.660 ns** |  **3.423 ns** |  **1.00** | **0.3548** |     **-** |     **-** |    **1488 B** |
| TrimGraphQLTypesNew |         [[Human!]!]! |    248.52 ns |  0.511 ns |  0.478 ns |  0.11 | 0.0076 |     - |     - |      32 B |
|                     |                      |              |           |           |       |        |       |       |           |
| **TrimGraphQLTypesOld** | **[[[[[(...)!!!!! [92]** | **23,615.13 ns** | **83.258 ns** | **69.524 ns** |  **1.00** | **4.5471** |     **-** |     **-** |   **19136 B** |
| TrimGraphQLTypesNew | [[[[[(...)!!!!! [92] |  1,531.74 ns | 29.858 ns | 34.385 ns |  0.06 |      - |     - |     - |         - |

### CODE REVIEW

File: src/GraphQL/GraphQL.csproj [hunk 1]
```diff
@@ -16,6 +16,7 @@
     <PackageReference Include="GraphQL-Parser" Version="5.3.3" />
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
     <PackageReference Include="System.ComponentModel.Annotations" Version="4.7.0" />
+    <PackageReference Include="System.Memory" Version="4.5.4" Condition="'$(TargetFramework)' == 'netstandard2.0'" />
```
Comment (sungam3r) [inline src/GraphQL/GraphQL.csproj:19]: I am absolutely sure that we will need this package in the near future. In addition, if we target the package also to netstandard2.1, then all the necessary APIs will be available out of the box, no code changes required.

Comment (sungam3r) [reply src/GraphQL/GraphQL.csproj:19]: ~ +800 KB dependencies but it's worth it

File: src/GraphQL/GraphQLExtensions.cs [hunk 2]
```diff
@@ -16,9 +16,30 @@ public static class GraphQLExtensions
...
+
+                int index = 0;
+                for (int i=0; i<name.Length; ++i)
+                {
+                    char c = name[i];
+                    if (!_bangs.Contains(c))
+                        buffer[index++] = c;
+                }
+
+                return index == name.Length ? name : buffer.Slice(0, index).ToString();
```
Comment (sungam3r) [inline src/GraphQL/GraphQLExtensions.cs:37]: 0 or 1 result string allocation

File: src/GraphQL/GraphQLExtensions.cs [hunk 3]
```diff
@@ -16,9 +16,30 @@ public static class GraphQLExtensions
...
+            }
+            else
+            {
+                return _trimPattern.Replace(name, string.Empty);
```
Comment (sungam3r) [inline src/GraphQL/GraphQLExtensions.cs:41]: I decided not to change this code branch at all. Hardly anyone has such long names.

**Comment by Shane32** [comment]
Did you not see that I removed this method entirely a few hours ago in PR #2144 .  It is not needed by the codebase whatsoever.

**Comment by Shane32** [comment]
I would like to see the speed comparison of this code to `name.Trim(_bangs)`.  I suspect it will be very close.

**Comment by sungam3r** [comment]
I did not. I look at the backlog of email.

**Comment by sungam3r** [comment]
You PR targets develop. Let's merge this PR in master first.

File: src/GraphQL/GraphQL.csproj (hunk 1 above)
Comment (Shane32) [reply src/GraphQL/GraphQL.csproj:19]: I just have a hard time thinking that this change is worth another dependency in the final version of v3 before we move to v4.  Isn't this code only used when building a schema?

For v4, I'd be happy to multi-target .net standard 2.0 and 2.1, adding system.buffers or similar MS-provided dependencies if we need to.

Comment (sungam3r) [reply src/GraphQL/GraphQL.csproj:19]: > Isn't this code only used when building a schema?

I see a lot of optimizations even without multi-targeting.

**Comment by sungam3r** [comment]
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.16299.726 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7 CPU 920 2.67GHz (Nehalem), 1 CPU, 8 logical and 4 physical cores
Frequency=2607421 Hz, Resolution=383.5207 ns, Timer=TSC
.NET Core SDK=5.0.101
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  DefaultJob : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT


```
|         Method |                 Name |       Mean |     Error |    StdDev |     Median | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |--------------------- |-----------:|----------:|----------:|-----------:|------:|--------:|-------:|------:|------:|----------:|
| **ToCamelCaseOld** |                     **** |   **6.120 ns** | **0.0157 ns** | **0.0131 ns** |   **6.118 ns** |  **1.00** |    **0.00** |      **-** |     **-** |     **-** |         **-** |
| ToCamelCaseNew |                      |  11.896 ns | 0.0360 ns | 0.0319 ns |  11.904 ns |  1.94 |    0.01 |      - |     - |     - |         - |
|                |                      |            |           |           |            |       |         |        |       |       |           |
| **ToCamelCaseOld** | **Loooo(...)ooong [48]** | **103.656 ns** | **0.3983 ns** | **0.3726 ns** | **103.536 ns** |  **1.00** |    **0.00** | **0.0631** |     **-** |     **-** |     **264 B** |
| ToCamelCaseNew | Loooo(...)ooong [48] |  70.383 ns | 0.2957 ns | 0.2766 ns |  70.389 ns |  0.68 |    0.00 | 0.0286 |     - |     - |     120 B |
|                |                      |            |           |           |            |       |         |        |       |       |           |
| **ToCamelCaseOld** |                **Short** |  **68.848 ns** | **0.1533 ns** | **0.1434 ns** |  **68.853 ns** |  **1.00** |    **0.00** | **0.0210** |     **-** |     **-** |      **88 B** |
| ToCamelCaseNew |                Short |  50.948 ns | 0.0616 ns | 0.0546 ns |  50.929 ns |  0.74 |    0.00 | 0.0076 |     - |     - |      32 B |
|                |                      |            |           |           |            |       |         |        |       |       |           |
| **ToCamelCaseOld** | **loooo(...)ooong [48]** |  **16.769 ns** | **0.0223 ns** | **0.0198 ns** |  **16.768 ns** |  **1.00** |    **0.00** |      **-** |     **-** |     **-** |         **-** |
| ToCamelCaseNew | loooo(...)ooong [48] |  16.809 ns | 0.0277 ns | 0.0232 ns |  16.809 ns |  1.00 |    0.00 |      - |     - |     - |         - |
|                |                      |            |           |           |            |       |         |        |       |       |           |
| **ToCamelCaseOld** |                **short** |  **16.763 ns** | **0.0239 ns** | **0.0212 ns** |  **16.762 ns** |  **1.00** |    **0.00** |      **-** |     **-** |     **-** |         **-** |
| ToCamelCaseNew |                short |  16.963 ns | 0.6755 ns | 1.9918 ns |  15.576 ns |  1.07 |    0.11 |      - |     - |     - |         - |

**Comment by Shane32** [comment]
> I would like to see the speed comparison of this code to `name.Trim(_bangs)`.

Did you try this yet?

**Comment by sungam3r** [comment]
In the case of strings for which the optimized conditional branch does not fire, the execution time is slightly longer, although the code is not even executed. I think this is because of JIT.

**Comment by sungam3r** [comment]
> Did you try this yet?

I did not understand.

**Comment by sungam3r** [comment]
I'm done for today, 3:20 am.

**Comment by Shane32** [comment]
> > Did you try this yet?
> 
> I did not understand.

Can you try comparing the TrimGraphQLTypes old and new against this:

```cs
        private static readonly char[] _chars = new[] { '[', ']', '!' };
        [Benchmark]
        public string ToCamelCaseAlt()
        {
            return Name.Trim(_chars);
        }
```

**Comment by sungam3r** [comment]
Indeed, the results are even better. This is due to the fact that `Trim` only strips off characters from the ends of the line, and does not go across the entire line, unlike the other two implementations.

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.16299.726 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7 CPU 920 2.67GHz (Nehalem), 1 CPU, 8 logical and 4 physical cores
Frequency=2607421 Hz, Resolution=383.5207 ns, Timer=TSC
.NET Core SDK=5.0.101
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  DefaultJob : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT


```
| Method |                 Name |         Mean |     Error |    StdDev |       Median | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------------------- |-------------:|----------:|----------:|-------------:|------:|-------:|------:|------:|----------:|
|    **Old** |                     **** |    **126.58 ns** |  **0.698 ns** |  **0.653 ns** |    **126.71 ns** |  **1.00** |      **-** |     **-** |     **-** |         **-** |
|    New |                      |     18.95 ns |  0.013 ns |  0.011 ns |     18.95 ns |  0.15 |      - |     - |     - |         - |
|    Alt |                      |     16.07 ns |  0.011 ns |  0.010 ns |     16.07 ns |  0.13 |      - |     - |     - |         - |
|        |                      |              |           |           |              |       |        |       |       |           |
|    **Old** |                **Human** |    **192.70 ns** |  **0.577 ns** |  **0.540 ns** |    **192.37 ns** |  **1.00** |      **-** |     **-** |     **-** |         **-** |
|    New |                Human |    101.22 ns |  0.017 ns |  0.013 ns |    101.21 ns |  0.53 |      - |     - |     - |         - |
|    Alt |                Human |     28.21 ns |  0.007 ns |  0.006 ns |     28.21 ns |  0.15 |      - |     - |     - |         - |
|        |                      |              |           |           |              |       |        |       |       |           |
|    **Old** |               **Human!** |    **491.95 ns** |  **0.637 ns** |  **0.565 ns** |    **492.03 ns** |  **1.00** | **0.0572** |     **-** |     **-** |     **240 B** |
|    New |               Human! |    133.11 ns |  0.089 ns |  0.083 ns |    133.11 ns |  0.27 | 0.0076 |     - |     - |      32 B |
|    Alt |               Human! |     45.12 ns |  0.161 ns |  0.151 ns |     45.04 ns |  0.09 | 0.0076 |     - |     - |      32 B |
|        |                      |              |           |           |              |       |        |       |       |           |
|    **Old** |              **[Human]** |    **774.36 ns** |  **1.151 ns** |  **1.020 ns** |    **774.01 ns** |  **1.00** | **0.1068** |     **-** |     **-** |     **448 B** |
|    New |              [Human] |    154.94 ns |  0.043 ns |  0.036 ns |    154.94 ns |  0.20 | 0.0076 |     - |     - |      32 B |
|    Alt |              [Human] |     47.20 ns |  1.058 ns |  2.387 ns |     45.48 ns |  0.06 | 0.0076 |     - |     - |      32 B |
|        |                      |              |           |           |              |       |        |       |       |           |
|    **Old** |             **[Human]!** |  **1,062.48 ns** |  **1.913 ns** |  **1.790 ns** |  **1,062.72 ns** |  **1.00** | **0.1564** |     **-** |     **-** |     **656 B** |
|    New |             [Human]! |    168.92 ns |  0.075 ns |  0.070 ns |    168.92 ns |  0.16 | 0.0076 |     - |     - |      32 B |
|    Alt |             [Human]! |     50.73 ns |  0.046 ns |  0.038 ns |     50.72 ns |  0.05 | 0.0076 |     - |     - |      32 B |
|        |                      |              |           |           |              |       |        |       |       |           |
|    **Old** |         **[[Human!]!]!** |  **2,120.14 ns** |  **1.554 ns** |  **1.454 ns** |  **2,119.83 ns** |  **1.00** | **0.3548** |     **-** |     **-** |    **1488 B** |
|    New |         [[Human!]!]! |    246.50 ns |  0.455 ns |  0.426 ns |    246.58 ns |  0.12 | 0.0076 |     - |     - |      32 B |
|    Alt |         [[Human!]!]! |     62.86 ns |  0.129 ns |  0.121 ns |     62.91 ns |  0.03 | 0.0076 |     - |     - |      32 B |
|        |                      |              |           |           |              |       |        |       |       |           |
|    **Old** | **[[[[[(...)!!!!! [92]** | **23,052.93 ns** | **50.256 ns** | **44.550 ns** | **23,069.84 ns** |  **1.00** | **4.5471** |     **-** |     **-** |   **19136 B** |
|    New | [[[[[(...)!!!!! [92] |  1,472.97 ns |  0.249 ns |  0.221 ns |  1,472.92 ns |  0.06 |      - |     - |     - |         - |
|    Alt | [[[[[(...)!!!!! [92] |    243.32 ns |  0.051 ns |  0.042 ns |    243.31 ns |  0.01 |      - |     - |     - |         - |

**Comment by sungam3r** [comment]
> I think this is because of JIT

And inlining. I also tried to rework `GetFriendlyName`: 
```c#
internal static string GetFriendlyName(this Type type)
        {
            if (type.GetGenericArguments().Length == 0) // fast return
                return type.Name;

            Span<char> buffer = stackalloc char[256];

            try
            {
                FillFriendlyNameRec(type, buffer, out int written);
                return buffer.Slice(0, written).ToString();
            }
            catch (ArgumentException)
            {
                return GetFriendlyNameOld(type);
            }
            catch (IndexOutOfRangeException)
            {
                return GetFriendlyNameOld(type);
            }

            static string GetFriendlyNameOld(Type type)
            {
                string friendlyName = type.Name;

                var genericArgs = type.GetGenericArguments();

                if (genericArgs.Length > 0)
                {
                    int iBacktick = friendlyName.IndexOf('`');
                    if (iBacktick > 0)
                    {
                        friendlyName = friendlyName.Remove(iBacktick);
                    }
                    friendlyName += "<";
                    for (int i = 0; i < genericArgs.Length; ++i)
                    {
                        string typeParamName = GetFriendlyNameOld(genericArgs[i]);
                        friendlyName += i == 0 ? typeParamName : "," + typeParamName;
                    }
                    friendlyName += ">";
                }

                return friendlyName;
            }
        }

        private static void FillFriendlyNameRec(Type type, Span<char> buffer, out int written)
        {
            string name = type.Name;

            var genericArgs = type.GetGenericArguments();

            if (genericArgs.Length == 0)
            {
                name.AsSpan().CopyTo(buffer);
                written = name.Length;
            }
            else
            {
                int iBacktick = name.IndexOf('`');
                written = 1;
                if (iBacktick > 0)
                {
                    name.AsSpan(0, iBacktick).CopyTo(buffer);
                    written += iBacktick;
                }
                buffer[iBacktick] = '<';
                buffer = buffer.Slice(written);

                for (int i = 0; i < genericArgs.Length; ++i)
                {
                    if (i > 0)
                    {
                        buffer[0] = ',';
                        written += 1;
                        buffer = buffer.Slice(1);
                    }
                    FillFriendlyNameRec(genericArgs[i], buffer, out int written2);
                    written += written2;
                    buffer = buffer.Slice(written2);
                }
                buffer[0] = '>';
                written += 1;
            }
        }
```

Test:
```c#
 [MemoryDiagnoser]
    public class GetFriendlyNameBenchmark
    {
        internal static string GetFriendlyNameOld(Type type)
        {
            string friendlyName = type.Name;

            var genericArgs = type.GetGenericArguments();

            if (genericArgs.Length > 0)
            {
                int iBacktick = friendlyName.IndexOf('`');
                if (iBacktick > 0)
                {
                    friendlyName = friendlyName.Remove(iBacktick);
                }
                friendlyName += "<";
                for (int i = 0; i < genericArgs.Length; ++i)
                {
                    string typeParamName = GetFriendlyNameOld(genericArgs[i]);
                    friendlyName += i == 0 ? typeParamName : "," + typeParamName;
                }
                friendlyName += ">";
            }

            return friendlyName;
        }

        [Benchmark(Baseline = true)]
        public string Old() => GetFriendlyNameOld(Type);

        [Benchmark]
        public string New() => Type.GetFriendlyName();

        [ParamsSource(nameof(Names))]
        public Type Type { get; set; }

        public IEnumerable<Type> Names => new[] { typeof(int), typeof(List<bool>), typeof(List<Dictionary<string, HashSet<DateTime>>>), typeof(List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<List<int>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>) };
    }
```
Results:
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.16299.726 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7 CPU 920 2.67GHz (Nehalem), 1 CPU, 8 logical and 4 physical cores
Frequency=2607421 Hz, Resolution=383.5207 ns, Timer=TSC
.NET Core SDK=5.0.101
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  DefaultJob : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT


```
| Method |                 Type |         Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------------------- |-------------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
|    **Old** | **Syste(...)lean] [49]** |    **374.31 ns** |  **1.725 ns** |  **1.614 ns** |  **1.00** |    **0.00** | **0.0458** |     **-** |     **-** |     **192 B** |
|    New | Syste(...)lean] [49] |    513.71 ns |  9.965 ns |  8.321 ns |  1.37 |    0.02 | 0.0267 |     - |     - |     112 B |
|        |                      |              |           |           |       |         |        |       |       |           |
|    **Old** |  **Syst(...)e]]] [143]** |  **1,151.44 ns** |  **2.947 ns** |  **2.757 ns** |  **1.00** |    **0.00** | **0.2346** |     **-** |     **-** |     **984 B** |
|    New |  Syst(...)e]]] [143] |  1,019.70 ns | 16.278 ns | 15.227 ns |  0.89 |    0.01 | 0.0591 |     - |     - |     248 B |
|        |                      |              |           |           |       |         |        |       |       |           |
|    **Old** | **Syst(...)]]]] [1307]** | **13,786.63 ns** | **22.116 ns** | **20.688 ns** |  **1.00** |    **0.00** | **5.4932** |     **-** |     **-** |   **22984 B** |
|    New | Syst(...)]]]] [1307] |  8,656.96 ns | 21.719 ns | 20.316 ns |  0.63 |    0.00 | 0.3967 |     - |     - |    1699 B |
|        |                      |              |           |           |       |         |        |       |       |           |
|    **Old** |         **System.Int32** |     **55.11 ns** |  **0.045 ns** |  **0.040 ns** |  **1.00** |    **0.00** |      **-** |     **-** |     **-** |         **-** |
|    New |         System.Int32 |     71.01 ns |  0.071 ns |  0.067 ns |  1.29 |    0.00 |      - |     - |     - |         - |


On "big data", where many generic parameters have to be processed, the new algorithm is undoubtedly better. But for simple types, it loses to the old one. Also, this method is only called to get the text of errors in exceptions. Therefore, I think that it is not worth changing it.

**Comment by sungam3r** [comment]
PR is ready for review.

**Review by Shane32 — COMMENTED** [review]
The changes look great.  But as it stands now, `FastChangeFirstLetter` is the only code requiring a dependency on System.Buffers.  I do not think that's worth the changes so late in the v3.x lifecycle.  Let's implement the change in TrimGraphQLTypes and change ToCamelCase in v4 adding the dependency there.

**Comment by Shane32** [comment]
Based on the tests you provided, the new `ToCamelCase` function, is 30 billionths of a second faster, at most, than the old function.  Even if you ran it 10,000 times it would still only be 0.3ms faster -- but that's only in a best-case scenario -- many of the metrics shows about even performance.  I suggest you run the introspection query configured as a transient and see if you can detect any measurable performance difference.  Again, I don't mind the change if we are adding System.Memory anyway, but we should not do so for v3, especially if there is no measurable performance difference when measuring some type of end-to-end test.

**Comment by sungam3r** [comment]
Done. `SchemaPrinter` is a good candidate for optimization.

**Review by Shane32 — APPROVED** [review]

**Pull request merged**